comp/def/lifecycle.go

```diff
@@ -0,0 +1,43 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package compDef
+
+import (
+	"context"
+)
+
+type lchFunc func(context.Context) error
+
+// Hook represents a function pair for component startup and shutdown
+type Hook struct {
+	OnStart lchFunc
+	OnStop  lchFunc
+}
+
+// HookStorage keeps track of Hooks, should not be used directly by client code
+type HookStorage struct {
+	hooks []Hook
+}
+
+// Lifecycle should be added to a component's requires struct if it wants to add hooks
+type Lifecycle struct {
+	storage *HookStorage
+}
```

Comment:
Having a constructor for the Lifecycle would help testing.

Also, an interface would be nice to allow the use of anything for the lifecycle (fx or other custom implementation).

Reply:
Changed to use an interface, but that leads to another problem, see my comment here: https://github.com/DataDog/datadog-agent/pull/24790/files#diff-0f33bf5aa71768f28c4ea77005ff4d1cbc65818e81e0038c314e87c19dc98f95R96